transports/quic/src/webtransport/connection.rs

```diff
@@ -0,0 +1,127 @@
+use h3_quinn::Connection as Http3Connection;
+use h3_webtransport::server::WebTransportSession;
+use bytes::Bytes;
+use crate::Error;
+use futures::{future::BoxFuture, FutureExt};
+use libp2p_core::muxing::{StreamMuxer, StreamMuxerEvent};
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+use h3::quic::BidiStream;
+use crate::webtransport::stream::Stream;
+
+/// State for a single opened WebTransport session.
+pub struct Connection {
+    /// Underlying connection.
+    session: WebTransportSession<Http3Connection, Bytes>,
+    /// Future for accepting a new incoming bidirectional stream.
+    incoming: Option<
+        BoxFuture<'static, (h3_webtransport::stream::SendStream<h3_quinn::SendStream<Bytes>, Bytes>,
+                            h3_webtransport::stream::RecvStream<h3_quinn::RecvStream, Bytes>)>,
+    >,
+    /// Future for opening a new outgoing bidirectional stream.
+    outgoing: Option<
+        BoxFuture<'static, (h3_webtransport::stream::SendStream<h3_quinn::SendStream<Bytes>, Bytes>,
+                            h3_webtransport::stream::RecvStream<h3_quinn::RecvStream, Bytes>)>,
+    >,
+    /// Future to wait for the connection to be closed.
+    closing: Option<BoxFuture<'static, h3::Error>>,
+}
+
+impl Connection {
+    /// Build a [`Connection`] from raw components.
+    ///
+    /// This function assumes that the [`quinn::Connection`] is completely fresh and none of
+    /// its methods has ever been called. Failure to comply might lead to logic errors and panics.
+    pub(crate) fn new(session: WebTransportSession<Http3Connection, Bytes>) -> Self {
+        Self {
+            session,
+            incoming: None,
+            outgoing: None,
+            closing: None,
+        }
+    }
+}
+
+impl StreamMuxer for Connection {
+    type Substream = Stream;
+    type Error = Error;
+
+    fn poll_inbound(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<Self::Substream, Self::Error>> {
+        let this = self.get_mut();
+
+        let incoming = this.incoming.get_or_insert_with(|| {
+            let accept = this.session.accept_bi();
```

Comment:
If you `Arc` the `WebTransportSession` then you can clone it here and fix the lifetime issues.